src/StreamJsonRpc.Tests/JsonRpcOverloadTests.cs

```diff
@@ -0,0 +1,253 @@
+        for (int iteration = 0; iteration < 10; iteration++)
+        {
+            using (var cts = new CancellationTokenSource())
+            {
+                this.clientStream.BeforeWrite = (stream, buffer, offset, count) =>
+                {
+                    // Cancel on the first write, when the header is being written but the content is not yet.
+                    if (!cts.IsCancellationRequested)
+                    {
+                        cts.Cancel();
+                    }
+                };
+
+                await Assert.ThrowsAsync<TaskCanceledException>(() => this.clientRpc.InvokeWithCancellationAsync<string>(nameof(Server.AsyncMethodWithCancellation), new[] { "a" }, cts.Token)).WithTimeout(UnexpectedTimeout);
+                this.clientStream.BeforeWrite = null;
+            }
+
+            // Verify that json rpc is still operational after cancellation.
+            // If the cancellation breaks the json rpc, like in https://github.com/Microsoft/vs-streamjsonrpc/issues/55, it will close the stream
+            // and cancel the request, resulting in unexpected OperationCancelledException thrown from the next InvokeAsync
+            string result = await this.clientRpc.InvokeAsync<string>(nameof(Server.ServerMethod), "a");
+            Assert.Equal("a!", result);
+        }
+    }
+
+    [Fact]
+    public async Task CancelMessageSentWhileAwaitingResponse()
+    {
+        using (var cts = new CancellationTokenSource())
+        {
+            var invokeTask = this.clientRpc.InvokeWithCancellationAsync<string>(nameof(Server.AsyncMethodWithCancellation), new[] { "a" }, cts.Token);
+            await this.server.ServerMethodReached.WaitAsync(this.TimeoutToken);
+            cts.Cancel();
+
+            // Ultimately, the server throws because it was canceled.
+            var ex = await Assert.ThrowsAnyAsync<OperationCanceledException>(() => invokeTask.WithTimeout(UnexpectedTimeout));
+#if !NET452
+            Assert.Equal(cts.Token, ex.CancellationToken);
+#endif
+        }
+
+        var result = await this.clientRpc.InvokeAsync<string>(nameof(Server.ServerMethod), "testing");
+        Assert.Equal("testing!", result);
+    }
+
+    [Fact]
+    public async Task InvokeWithCancellationAsync_CanCallCancellableMethodWithNoArgs()
+    {
+        Assert.Equal(5, await this.clientRpc.InvokeWithCancellationAsync<int>(nameof(Server.AsyncMethodWithCancellationAndNoArgs)));
+
+        using (var cts = new CancellationTokenSource())
+        {
+            Task<int> resultTask = this.clientRpc.InvokeWithCancellationAsync<int>(nameof(Server.AsyncMethodWithCancellationAndNoArgs), cancellationToken: cts.Token);
+            cts.Cancel();
+            try
+            {
+                int result = await resultTask;
+                Assert.Equal(5, result);
+            }
+            catch (OperationCanceledException)
+            {
+                // this is also an acceptable result.
```

Comment:
If this is the exception sent by the server (iirc how this works), should we check for the integrity of the exception? Making sure it has the correct content, stack trace, and any other information that would be helpful to the user.